### ISSUE TITLE
Bash: pressing tab to select entry in reverse-i-search freezes the terminal

### ISSUE THREAD

**Issue by derekthesnake** [microsoft/terminal#14695]
### Windows Terminal version

1.15.3465.0

### Windows build number

Version 10.0.19044.2364

### Other Software

GNU bash, version 5.0.17(1)-release (x86_64-pc-linux-gnu)

### Steps to reproduce

Control + Shift + R to enter reverse-i-search, type to bring up a command then press TAB to try to select it.

### Expected Behavior

The command should be put at the prompt and the terminal should remain interactive.

### Actual Behavior

The command does get placed at the prompt but the terminal becomes unresponsive. Control-C is the only input that registers to the screen, but it doesn't unfreeze the terminal. All other keypresses seem to have no effect, although the blinking cursor does pause blinking when other keys are pressed so it seems that they are registered to some degree.

**Comment by derekthesnake** [comment]
Oh, yep, sorry forgot to include WSL. 

Hmm... I'll do some more investigating about the tab completion. Interesting behavior then, on my Linux hardware it wasn't triggering tab completion, just selecting the command. I'll look into the versions of bash I suppose.

**Comment by DHowett** [comment]
qq: If you were to `export PATH=/bin:/sbin:/usr/bin:/usr/sbin` before you do a `bck-i-search` + <kbd>Tab</kbd>, does it get better?

**Comment by JL102** [comment]
Hi @DHowett I'm experiencing the same problem. Do you have the ability to reopen this issue?

I ran `export PATH=/bin:/sbin:/usr/bin:/usr/sbin` before doing reverse-i-search and yes, it gets better. 
I don't think it's entirely the fault of system32. I copied my entire PATH and removed `/mnt/c/Windows/System32` from it, and I still got a long delay (I have a loooot of folders on my system PATH on Windows, so there's still a lot of items there).

I don't understand, though - Why would having a lot of items on your PATH affect `reverse-i-search`? Like, normal tab completion doesn't have the same delay. If I type `explo` and press tab, it autocompletes to `explorer` within about 1 second and then if I type out `explorer.e` and press tab again, it autocompletes to `explorer.exe` again within about 1 second. And actually _running_ `reverse-i-search` seems to have no delay at all, i.e. when I type it searches through my bash history really quickly. So what's it doing in the background after I press tab to fill out the whole command?

Thanks!

**Comment by JL102** [comment]
> I don't believe that Tab is the canonical way to select a result from bck-i-search, so I believe that you're triggering tab completion instead of accepting the result.

What **is** the normal way to select a result from `reverse-i-search`?

**Comment by DHowett** [comment]
Hey @JL102! Sorry to leave you on read for a couple days.

So, what I _believe_ is happening is that bash is scanning every directory in `$PATH` for commands whose names match the input up to the cursor. When there's nothing, everything matches. When there's something, very few things match. Now that's not usually a problem of course... but WSL is an interesting beast.

`/mnt/c` is a 9p network share that talks to a userland process on the Windows side. That's not really that slow either, but it's definitely much slower than direct VHD access (like `/`.) I think enumerating a directory over 9p is pretty slow... but it also gets a bit worse when (again, I think!) bash has to `stat()` _every file_ to determine whether it's executable (!). `stat` on Windows is terrifically slow.

(I don't know if `stat` info comes back with `readdir`... but if it does, then it's just the Windows P9 server doing the slow statting rather than bash :smile:)

> What **is** the normal way to select a result from `reverse-i-search`?

You know, I never thought about this! I always use the right arrow, which breaks out of bck-i-search and moves the cursor simultaneously. :shrug:

**Comment by JL102** [comment]
@DHowett Thanks for the explanation! Really sucks that it does this, but your answer seems to make sense.

> So, what I believe is happening is that bash is scanning every directory in $PATH for commands whose names match the input up to the cursor. When there's nothing, everything matches. When there's something, very few things match. Now that's not usually a problem of course... but WSL is an interesting beast.

Hmm. The explanation makes sense for what's causing it to take so long, but I still don't get why Bash is (probably) triggering tab autocompletion when I exit `reverse-i-search`. When exiting `reverse-i-search`, the command is fully populated, so I don't see any need for the tab autocompletion to run. Do you think it may be a bug in Bash, or is there a potential reason for it to behave this way?

**Comment by JosefSaltz** [comment]
I am still having this issue as of end of February 2024.

**Comment by JL102** [comment]
FYI, I've been working around this by pressing `End` on my keyboard instead of tab.

**Comment by JohnRDOrazio** [comment]
I've been having the same problem in a WSL terminal, and just stumbled upon this issue which explains what is happening. I spontaneously used <kbd>Tab</kbd> to select the current result, I guess I should use the right arrow or <kbd>End</kbd> from now on.

**Comment by dwymark-celestron** [comment]
This issue is still reproducing on Ubuntu WSL in Windows 11. The ticket ought to be reopened.

**Comment by DHowett** [comment]
> This issue is still reproducing on Ubuntu WSL in Windows 11. The ticket ought to be reopened.

@dwymark-celestron is the explanation above (copied here below) unsuitable?

> So, what I _believe_ is happening is that bash is scanning every directory in `$PATH` for commands whose names match the input up to the cursor. When there's nothing, everything matches. When there's something, very few things match. Now that's not usually a problem of course... but WSL is an interesting beast.
> 
> `/mnt/c` is a 9p network share that talks to a userland process on the Windows side. That's not really that slow either, but it's definitely much slower than direct VHD access (like `/`.) I think enumerating a directory over 9p is pretty slow... but it also gets a bit worse when (again, I think!) bash has to `stat()` _every file_ to determine whether it's executable (!). `stat` on Windows is terrifically slow.
> 
> (I don't know if `stat` info comes back with `readdir`... but if it does, then it's just the Windows P9 server doing the slow statting rather than bash 😄)
> 
> > What **is** the normal way to select a result from `reverse-i-search`?
> 
> You know, I never thought about this! I always use the right arrow, which breaks out of bck-i-search and moves the cursor simultaneously. 🤷

The TL;DR is: Windows Terminal and the Windows Console Subsystem are not responsible for this behavior. Your shell is.

**Comment by JohnRDOrazio** [comment]
Ever since I came across the same situation and after reading this thread, I started using the right arrow key instead of the Tab key. The Tab key seems more intuitive for an "autocomplete" feature. 

I believe this issue, given the situation, is a "can't fix", and the solution is just to get used to using the right arrow rather than the Tab key.

**Comment by dwymark-celestron** [comment]
@DHowett yes, I agree with that description of what is happening. I've dug into the issue deeper since I commented last, and I have a suggestion for anyone else who has a deeply ingrained habit of pressing TAB for completions.

You can completely prevent the TAB completion search from happening in the isearch context with a simple change to `readline`'s configuration `.inputrc`. First, open or create your `.inputrc` file in your home directory:
```shell
nano ~/.inputrc
```

Add the line below to override the search terminators list to include TAB:
```
set isearch-terminators "\033\012\011"
```

Note that the default isearch-terminators are "\033\012", which is to say ESC and Ctrl+J.

This configuration change ensures that bash **immediately** terminates the interactive search whenever the TAB key is pressed (tab being `\011` in octal). This is different from the default behavior, which is for the TAB key to be detected as a control character in this branch in `lib/readline/isearch.c` in the bash source:
```C
  if (MB_CUR_MAX > 1 && rl_byte_oriented == 0)
    {
      if (cxt->lastc >= 0 && (cxt->mb[0] && cxt->mb[1] == '\0') && ENDSRCH_CHAR (cxt->lastc))
	{
	  /* This sets rl_pending_input to LASTC; it will be picked up the next
	     time rl_read_key is called. */
	  rl_execute_next (cxt->lastc);
	  return (0);
	}
    }
```

The `rl_execute_next (cxt->lastc);` line is what ensures that the control character (TAB in this case) causes any associated behaviors to be executed as usual after the isearch terminates. With the addition of TAB as an isearch-terminator, this branch is never hit, and instead isearch terminates as below:
```C
  /* The characters in isearch_terminators (set from the user-settable
     variable isearch-terminators) are used to terminate the search but
     not subsequently execute the character as a command.  The default
     value is "\033\012" (ESC and C-J). */
  if (cxt->lastc > 0 && strchr (cxt->search_terminators, cxt->lastc))
    {
      /* ESC still terminates the search, but if there is pending
	 input or if input arrives within 0.1 seconds (on systems
	 with select(2)) it is used as a prefix character
	 with rl_execute_next.  WATCH OUT FOR THIS!  This is intended
	 to allow the arrow keys to be used like ^F and ^B are used
	 to terminate the search and execute the movement command.
	 XXX - since _rl_input_available depends on the application-
	 settable keyboard timeout value, this could alternatively
	 use _rl_input_queued(100000) */
      if (cxt->lastc == ESC && (_rl_pushed_input_available () || _rl_input_available ()))
	rl_execute_next (ESC);
      return (0);
```

Hope this is helpful for someone.

**Comment by derekthesnake** [comment]
> What **is** the normal way to select a result from `reverse-i-search`?

@DHowett @JL102 I finally have a definitive answer to this. readline is what handles the reverse searching, and from the man page: 

```
isearch-terminators (‘‘C-[ C-J'')
              The  string  of  characters  that  should  terminate an incremental
              search without subsequently executing the character as  a  command.
              If this variable has not been given a value, the characters ESC and
              C-J will terminate an incremental search.
```

I just read the readline man page and immediately thought about this issue. Not that it helps with the WSL slowness, but it seems like DHowett has gotten to the bottom of that.

**Comment by dwymark-celestron** [comment]
It does solve the problem to add tab to the inputrc, as I suggested above. I daily drive this configuration on WSL :)
